src/drivers/irlock/irlock.cpp

```diff
@@ -0,0 +1,410 @@
+ *   Author: Michael Landes
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/*
+ * @file irlock.cpp
+ *
+ * Driver for an IR-Lock and Pixy vision sensor connected via I2C.
+ *
+ */
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include <drivers/boards/px4fmu-v2/board_config.h>
+#include <drivers/device/i2c.h>
+#include <drivers/device/ringbuffer.h>
+#include <drivers/drv_irlock.h>
+#include <drivers/drv_hrt.h>
+
+#include <nuttx/clock.h>
+#include <nuttx/wqueue.h>
+#include <systemlib/err.h>
+
+/* Configuration Constants */
+#define IRLOCK_I2C_BUS			PX4_I2C_BUS_EXPANSION
+#define IRLOCK_I2C_ADDRESS		0x65 //* 7-bit address (non shifted)
+#define IRLOCK_CONVERSION_INTERVAL_US	20000U /* us = 20ms = 50Hz */
+
+#define IRLOCK_SYNC			0xAA55
+#define IRLOCK_RESYNC		0x5500
+#define IRLOCK_ADJUST		0xAA
+
+#ifndef CONFIG_SCHED_WORKQUEUE
+# error This requires CONFIG_SCHED_WORKQUEUE.
+#endif
+
+class IRLOCK : public device::I2C
+{
+public:
+	IRLOCK(int bus = IRLOCK_I2C_BUS, int address = IRLOCK_I2C_ADDRESS);
+	virtual ~IRLOCK();
+
+	virtual int init();
+	virtual int probe();
+	virtual int info();
+
+	virtual ssize_t read(struct file *filp, char *buffer, size_t buflen);
+
+private:
+
+	/* start periodic reads from sensor */
+	void start();
+
+	/* stop periodic reads from sensor */
+	void stop();
```

Comment:
Code style: start() and stop() should be on the same column as cycle_trampoline()